File: draw_figures/plotActions.m
function plotActions(index, actions, color, xRange)
temp = selectOneAction(actions, 1);
plot(index, temp, 'Color', color(1,:), 'LineWidth', 15);
hold on;
temp = selectOneAction(actions, 2);
plot(index, temp, 'Color', color(2,:), 'LineWidth', 15);
temp = selectOneAction(actions, 3);
plot(index, temp, 'Color', color(3,:), 'LineWidth', 15);
hold off
ylim([0.8 1.2])
xlim(xRange)
box off
set(gca, 'YColor', 'white')
set(gca, 'XColor', 'white')
set(gca, 'TickLength', [0 0])
end

function x = selectOneAction(actions, a)
x = 1*(actions==a); % pick out one actions
% change 010 into 111 to line up (If not, 010 can not be seen)
singleIdxes = strfind(x, [0 1 0]);
for i=1:length(singleIdxes)
    x(singleIdxes(i):(singleIdxes(i)+2)) = [1 1 1];
end
% do not plot 0 values
x(x==0)=nan;
end

File: Python_Ver/draw_figures/plotActions.py
import numpy as np
import matplotlib.pyplot as plt

def selectOneAction(actions, a):
    """
    Pick out one action and process it for visualization.

    Parameters:
    actions : array-like
        The sequence of actions.
    a : int
        The action to be selected.

    Returns:
    x : numpy array
        Processed action sequence for plotting.
    """
    x = (actions == a).astype(float)  # Pick out the specific action

    # Convert 010 into 111 to make it visible in the plot
    single_indexes = np.where((x[:-2] == 0) & (x[1:-1] == 1) & (x[2:] == 0))[0]
    for idx in single_indexes:
        x[idx:idx+3] = [1, 1, 1]

    # Convert 0 values to NaN to avoid plotting them
    x[x == 0] = np.nan
    return x

def plotActions(ax, index, actions, color, xRange):
    """
    Plot action sequences with different colors.

    Parameters:
    index : array-like
        The x-axis indices.
    actions : array-like
        The sequence of actions.
    color : array-like
        A list of colors for different actions.
    xRange : tuple
        The x-axis range for the plot.
    """
    
    for i in range(1, 4):  # Actions 1, 2, and 3
        plot_actions = selectOneAction(actions, i)
        ax.plot(index, plot_actions, color=color[i-1], linewidth=15, solid_capstyle='butt')

    ax.set_ylim(0.8, 1.2)
    ax.set_xlim(xRange)
    plt.box(on=False)
    ax.tick_params(axis='y', colors='white', length=0)
    ax.tick_params(axis='x', colors='white', length=0)
